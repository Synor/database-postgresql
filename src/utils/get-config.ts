import { SynorError } from '@synor/core'
import { ConnectionString } from 'connection-string'
import { readFileSync } from 'fs'
import { resolve as resolvePath } from 'path'

type ConnectionConfig = import('pg').ConnectionConfig
type TLSConnectionOptions = import('tls').ConnectionOptions

type SSLConfig = Pick<
  TLSConnectionOptions,
  'passphrase' | 'rejectUnauthorized' | 'secureOptions' | 'ca' | 'key' | 'cert'
>

type SSLParams = Pick<
  SSLConfig,
  'passphrase' | 'rejectUnauthorized' | 'secureOptions'
> & {
  ca?: string
  key?: string
  cert?: string
}

type PostgreSQLDatabaseConfig = Required<Pick<ConnectionConfig, 'database'>> &
  Pick<ConnectionConfig, 'host' | 'port' | 'user' | 'password' | 'ssl'> & {
    application_name?: string
  }

type PostgreSQLEngineConfig = {
  migrationRecordTable: string
  schema: string
}

export function getConfig(
  uri: string
): [PostgreSQLDatabaseConfig, PostgreSQLEngineConfig] {
  try {
    const {
      protocol,
      hostname: host,
      port,
      user,
      password,
      path,
      params = {}
    } = new ConnectionString(uri, {
      params: {
        application_name: '@synor/database-postgresql',
        schema: 'public',
        synor_migration_record_table: 'synor_migration_record'
      }
    })

    if (!protocol) {
      throw new Error(`[URI] missing: protocol!`)
    }

    if (protocol !== 'postgresql') {
      throw new Error(`[URI] unsupported: protocol(${protocol})!`)
    }

    const database = path && path[0]

    if (!database) {
      throw new Error('[URI] missing: database!')
    }

    const sslParams: boolean | SSLParams = JSON.parse(params.ssl || false)

    let ssl: boolean | SSLConfig

    if (typeof sslParams === 'boolean') {
      ssl = sslParams
    } else {
      ssl = {
        passphrase: sslParams.passphrase,
        rejectUnauthorized: sslParams.rejectUnauthorized,
        secureOptions: sslParams.secureOptions
      }

      if (sslParams.ca) {
        ssl.ca = readFileSync(resolvePath(sslParams.ca))
      }
      if (sslParams.cert) {
        ssl.cert = readFileSync(resolvePath(sslParams.cert))
      }
      if (sslParams.key) {
        ssl.key = readFileSync(resolvePath(sslParams.key))
      }
    }

    const databaseConfig: PostgreSQLDatabaseConfig = {
      database,
      host,
      port,
      user,
      password,
      ssl,
      application_name: params.application_name
    }

    const engineConfig: PostgreSQLEngineConfig = {
      migrationRecordTable: params.synor_migration_record_table,
      schema: params.schema
    }

    return [databaseConfig, engineConfig]
  } catch (error) {
    throw new SynorError('Invalid DatabaseURI', error)
  }
}
