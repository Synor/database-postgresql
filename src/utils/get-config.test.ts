import fs from 'fs'
import { getConfig } from './get-config'

describe('utils:getConfig', () => {
  let uri: Parameters<typeof getConfig>[0]

  beforeEach(() => {
    uri = 'postgresql://postgres:postgres@127.0.0.1:5432/synor'
  })

  test('accepts postgresql uri', () => {
    expect(getConfig(uri)).toMatchSnapshot()
  })

  test.each(['postgres:', 'mysql:'])(
    `throws if uri protocol is not 'postgresql:'`,
    protocol => {
      uri = uri.replace('postgresql:', protocol)
      expect(() => getConfig(uri)).toThrow()
    }
  )

  test('throws if protocol is missing', () => {
    uri = uri.replace('postgresql:', '')
    expect(() => getConfig(uri)).toThrow()
  })

  test('throws if database is missing', () => {
    uri = uri.replace('/synor', '')
    expect(() => getConfig(uri)).toThrow()
  })

  test('throws if uri is malformed', () => {
    uri = 'postgresql://@ _ @/synor'
    expect(() => getConfig(uri)).toThrow()
  })

  test('accepts custom migration record table name', () => {
    const tableName = 'migration_history'
    uri = `${uri}?synor_migration_record_table=${tableName}`
    expect(getConfig(uri)[1].migrationRecordTable).toBe(tableName)
  })

  test('accepts schema name', () => {
    const schemaName = 'jest'
    uri = `${uri}?schema=${schemaName}`
    expect(getConfig(uri)[1].schema).toBe(schemaName)
  })

  describe('ssl config', () => {
    beforeEach(() => {
      uri = 'postgresql://postgres:postgres@127.0.0.1:5432/synor'
    })

    test('defaults to false', () => {
      expect(getConfig(uri)[0].ssl).toBe(false)
    })

    test.each([false, true])('accepts boolean: %s', ssl => {
      uri = `${uri}?ssl=${JSON.stringify(ssl)}`
      expect(getConfig(uri)[0].ssl).toBe(ssl)
    })

    test('accepts uri encoded stringified json', () => {
      const ssl = { rejectUnauthorized: false }
      uri = `${uri}?ssl=${encodeURIComponent(JSON.stringify(ssl))}`
      expect(getConfig(uri)[0].ssl).toMatchObject(ssl)
    })

    test.each([
      ['ca', 'CA'],
      ['cert', 'CERT'],
      ['key', 'KEY']
    ])(`reads ssl.%s file content`, (key, content) => {
      jest
        .spyOn(fs, 'readFileSync')
        .mockImplementationOnce((v: any) => Buffer.from(`CONTENT:${v}`))

      const ssl = { [key]: content }
      uri = `${uri}?ssl=${encodeURIComponent(JSON.stringify(ssl))}`
      expect(
        ((getConfig(uri)[0].ssl as any)[key] as Buffer).toString()
      ).toMatch(new RegExp(`CONTENT:.+${content}`))
    })
  })
})
