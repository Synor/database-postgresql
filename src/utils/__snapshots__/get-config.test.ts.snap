// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils:getConfig accepts postgresql uri 1`] = `
Array [
  Object {
    "application_name": "@synor/database-postgresql",
    "database": "synor",
    "host": "127.0.0.1",
    "password": "postgres",
    "port": 5432,
    "ssl": false,
    "user": "postgres",
  },
  Object {
    "migrationRecordTable": "synor_migration_record",
    "schema": "public",
  },
]
`;
